### DISCUSSION TITLE
autocommands

### DISCUSSION THREAD

**Discussion by Docbroke** [leo-arch/clifm#64] [Ideas]
for example, when user enters Downloads directory, change sorting method to ctime.
may be something like in config file
```
autocmd DirEnter *Downloads* SetLocal sort ctime
```

**Comment by leo-arch** [comment]
Never thought about this. But I guess it could be a useful feature, I like it.

So, the idea would be to be able to modify settings (and not only the sorting method) on a per directory basis, isn't? I mean, you should also be able to switch hidden files, folders first, files counter (this could be very useful if you know the directory is used to mount a remote filesystem, which is usually quite slower), and so on.

I need time to think of a good implementation. Added to the TODO list.

I you have any suggestion about it, that would be great.

Btw, do you know of any file manager implementing this feature? Some model to play around is always useful.

**Reply by Docbroke** [reply]
This feature is available in vifm, the config code I have used as an example is straight from vifm. Original ideas I think comes from vim autocommands.

**Reply by leo-arch** [reply]
Good to know. I'll take a look at it.

**Reply by leo-arch** [reply]
A first proof of concept for the autocommands function is ready and working. Expect news in a few days.

**Comment by leo-arch** [comment]
The first version of  the `autocommands` function is ready. This is how it works:

Add a line preceded by the `autocmd` keyword to the config file. The general syntax is: `autocmd PATTERN cmd,cmd,cmd`

`PATTERN` is a glob expression to match directory names. If no glob metacharacter is provided, the string will be compared as is with the current working directory. To invert the meaning of a pattern, prepend an exclamation mark: `!PATTERN`

`PATTERN` is followed by a comma separated list of commands:

* `!CMD`: The exclamation mark allows you to run shell commands
* The following codes are used to control CliFM's files list:

| Code | Description | Example |
| --- | --- | -- |
| `cs` | Color scheme | `cs=amber-256` |
| `fc` | Files counter | `fc=0` |
| `hf` | Hidden files | `hf=0` |
| `lm` | Light mode | `lm=1` |
| `lv` | Long/detail view | `lm=0` |
| `mf` | Max files | `mf=100` |
| `mn` | Max file name length | `mn=20` |
| `pg` | Pager | `pg=0` |
| `st` | Sort method | `st=5` |

A few example lines:

```sh
#autocmd !/media/* lm=1,fc=0,lv=0,cs=cga
autocmd /usr/bin mf=100,pg=1
autocmd /home/_leo08 hf=0,cs=amber-256,lv=1
autocmd /home/_leo08/Documents fc=0
autocmd /home/_leo08/Downloads mn=20,!~/.config/clifm/plugins/fzfnav.sh
autocmd / lv=1,fc=0,cs=solarized,st=5
```

**Reply by Docbroke** [reply]
I am getting build failed
```
src/autocmds.c: In function 'check_autocmds':
src/autocmds.c:68:21: error: 'GLOB_NOMAGIC' undeclared (first use in this function); did you mean 'GLOB_NOMATCH'?
   68 |   int ret = glob(p, GLOB_NOMAGIC | GLOB_TILDE
      |                     ^~~~~~~~~~~~
      |                     GLOB_NOMATCH
src/autocmds.c:68:21: note: each undeclared identifier is reported only once for each function it appears in
src/autocmds.c:69:20: error: 'GLOB_ONLYDIR' undeclared (first use in this function); did you mean 'IN_ONLYDIR'?
   69 |     | GLOB_BRACE | GLOB_ONLYDIR, NULL, &g);
      |                    ^~~~~~~~~~~~
      |                    IN_ONLYDIR
```

**Reply by leo-arch** [reply]
Try now.

**Reply by Docbroke** [reply]
compiled successfully.
EDIT: working fine

**Reply by leo-arch** [reply]
Great. Please report if you find anything unexpected.

**Reply by Docbroke** [reply]
How to catch all subdirectories in `PATTERN` ?
I am trying to  use `fc=0` in my remote mount directory using
```
autocmd ~/mount/** fc=0
```
but this doesn't seem to apply to all subdirectories.

**Reply by leo-arch** [reply]
Just one asterisk should do the trick:

```sh
autocmd ~/mount/* fc=0
```

If you want to include the ` mount` directory as well, write a separate line:

```sh
autocmd ~/mount/* fc=0
autocmd ~/mount fc=0
```
I should really add support for the double asterisk to include not only subdirectories, but the parent directory as well. Added to my TODO list.

**Reply by Docbroke** [reply]
I have tried both one and two asterix, but it is working only one level below the mount directory not further down.

**Reply by leo-arch** [reply]
I'll take a look at it.

**Reply by leo-arch** [reply]
Have you tried this:

```sh
autocmd ~/mount/* fc=0
autocmd ~/mount fc=0
```

Works fine for me.

**Reply by leo-arch** [reply]
Nope, you're right. I'll fix it.

**Reply by leo-arch** [reply]
It should be working now: use the double asterisk to recursively mark directories, both parent and all subdirs:

```sh
autocmd ~/mount/** fc=0
```

**Reply by leo-arch** [reply]
Still not working with tilde. Working on it.

**Reply by leo-arch** [reply]
Try now.

**Reply by Docbroke** [reply]
This is perfect. Thanks.

I was also testing `mf` command. I have one suggestion here.
`mf 50` lists first 50 files, but there is no way to see remaining files. Can we have `mf next|n` and `mf prev|p`  commands to see next or previous 50 files (when mf=50). This will be very useful when directory have large number of files and terminal doesn't support scrolling. This will be more useful compared `pager`, which lists files only one time.

**Reply by leo-arch** [reply]
You're welcome. Thank you for all your suggestions and testings; your feedback is very useful.

As to the `mf` command, I've considered that myself, but never implemented it. Added to my TODO list. It shouldn't be hard to implement, but one never knows.

**Comment by Docbroke** [comment]
Kindly allow `reverse sort` with autocommands. `st=n` has no reverse sort option.
This is useful for sorting by time, as normal sort lists latest file at the bottom.

**Reply by leo-arch** [reply]
Done.

EDIT: Use `sr=1` to set reverse sorting.
